### ISSUE TITLE
S3 permissions errors using Cromwell 37+

### ISSUE THREAD

**Issue by kshakir** [broadinstitute/cromwell#4740]
### Possible Workaround

Try using the `default` AWS auth scheme along with the [AWS default credentials](https://docs.aws.amazon.com/sdk-for-java/v2/developer-guide/credentials.html#credentials-default) setup on the system / environment.

### Background

Starting in cromwell 37, the AWS S3 SDK was upgraded from `2.0.0-preview-9` to `2.3.9`. Along with this upgrade several cromwell calls to the S3 SDK were updated to match changes in the library.

Post upgrade, the existing Cromwell AWS CI tests continue to pass, however there have been reports of permissions problems or other errors.
- https://gatkforums.broadinstitute.org/firecloud/discussion/comment/56245/#Comment_56245
- https://github.com/broadinstitute/cromwell/issues/4541
- https://github.com/broadinstitute/cromwell/issues/4686
- https://github.com/broadinstitute/cromwell/issues/4731

Because the CI tests are passing, and they use default credentials, it is possible that each of these issues may be also be worked around by using the [AWS default credentials](https://docs.aws.amazon.com/sdk-for-java/v2/developer-guide/credentials.html#credentials-default) scheme.

It is unclear how the very similar SDK calls worked with `2.0.0-preview-9` and not `2.3.9`. However the fix might include passing in credentials via the `Map env` / `Properties props`.

| Type                                         | Cromwell copy targeting `2.3.9`                                                                                                                                                      | "Original" targeting `2.0.0-preview-9`                                                                                                                                               |
|----------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| (Attempted) setting of key/secret from props | [cromwell-37/AmazonS3Factory.java](https://github.com/broadinstitute/cromwell/blob/37/filesystems/s3/src/main/java/org/lerch/s3fs/AmazonS3Factory.java#L81)          | [elerch-sdk2/AmazonS3Factory.java](https://github.com/elerch/Amazon-S3-FileSystem-NIO2/blob/sdk2/src/main/java/org/lerch/s3fs/AmazonS3Factory.java#L82)               |
| Region not set by props                      | [cromwell-37/AmazonS3Factory.java](https://github.com/broadinstitute/cromwell/blob/37/filesystems/s3/src/main/java/org/lerch/s3fs/AmazonS3Factory.java#L59)          | [elerch-sdk2/AmazonS3Factory.java](https://github.com/elerch/Amazon-S3-FileSystem-NIO2/blob/sdk2/src/main/java/org/lerch/s3fs/AmazonS3Factory.java#L57)               |
| (Attempted) props read from env              | [cromwell-37/AmazonS3Factory.java](https://github.com/broadinstitute/cromwell/blob/37/filesystems/s3/src/main/java/org/lerch/s3fs/S3FileSystemProvider.java#L95-L96) | [elerch-sdk2/S3FileSystemProvider.java](https://github.com/elerch/Amazon-S3-FileSystem-NIO2/blob/sdk2/src/main/java/org/lerch/s3fs/S3FileSystemProvider.java#L94-L95) |

### Acceptance Criteria

The existing four centaur tests (two run in Travis, two nightly in Jenkins) should pass using non-default credentials:
- Switch the `aws_application.conf` to use auth scheme `custom_keys` in a rendered template with secrets embedded
- Remove the files `aws_credentials.ctmpl` and `aws_config`
- Update `testCentaurAws.sh` removing `AWS_SHARED_CREDENTIALS_FILE` and `AWS_CONFIG_FILE` exports
  - **Maybe**: It isn't clear that installing the `awscli` should be required. Remove this from the `testCentaurAws.sh` and see if the tests still work using just the embedded Java S3 SDK.
- Update `.gitignore` replacing `aws_credentials` with `aws_application.conf`
- **Maybe**: Add a second backend that **does** use some form of default authorization
  - If adding this test, please ensure that the non-default credentials are NOT accidentally falling-back to the default credentials

### TL;DR

**It's unclear how cromwell calls to S3 broke in 37+, but CI testing with non-default credentials should hopefully reproduce reported errors.**

**Comment by AlekseiLitkovetc** [comment]
Hello!

I have recently started to work with the Cromwell project. Could you please clarify "Acceptance criteria" section? I would like to know which two tests should pass with non-default credentials.

Also, could you please prompt where I can find a link to Jenkins?

Thank you in advance!

**Comment by kshakir** [comment]
Hi @AlekseiLitkovetc,

>  I would like to know which two tests should pass with non-default credentials.

Back in March when I filed the ticket we were only [including tests for four](https://github.com/broadinstitute/cromwell/blob/38/src/ci/bin/testCentaurAws.sh#L24-L27) of our dozens of Centaur tests. After the AWS hackathon, the test coverage expanded to only [exclude a few](https://github.com/broadinstitute/cromwell/blob/39/src/ci/bin/testCentaurAws.sh#L25-L53) remaining tests.

So, the original minimum A/C was to see these tests passing with non-default creds:
- [hello](https://github.com/broadinstitute/cromwell/blob/42/centaur/src/main/resources/standardTestCases/hello.test)
- [long_cmd](https://github.com/broadinstitute/cromwell/blob/42/centaur/src/main/resources/standardTestCases/long_cmd.test)
- [haplotypecaller.aws](https://github.com/broadinstitute/cromwell/blob/42/centaur/src/main/resources/integrationTestCases/haplotypecaller.aws.test)
- [singlesample.aws](https://github.com/broadinstitute/cromwell/blob/42/centaur/src/main/resources/integrationTestCases/singlesample.aws.test)

> could you please prompt where I can find a link to Jenkins?

Our Jenkins servers are only internally accessible to Broad employees because of Jenkins continued problems with security. Perhaps one day we'll move off Travis to another public CI-as-a-Service that will run tests longer than 180 minutes, such as CircleCI or Google Cloud Build. Then we could migrate all tests to one place.

**Comment by geoffjentry** [comment]
It's worth noting that this is not necessary to get a working Cromwell-on-AWS set up. It's only necessary if one wants to use the default login from the CLI

**Comment by AlekseiLitkovetc** [comment]
@kshakir, thank you for your detailed explanation.

If you do not mind, could you please shed some light on the one more thing.
I have a little bit messy with the definition of what the default credentials are.
Does it mean that the credentials, which Travis and Jenkins use, is default credentials and the purpose of this task is to set other ones and check that four tests were successfully passed? In my point of view, it looks insecure to pass my keys and AWS environments for CI testing. This is why I think that I missed something.

Also, in case if I will suggest a fix for this task I will not be able to check the results of CI testing by myself cause I do not have access to the Jankins. Is it possible to get access in read-only mode or I can communicate with someone who can provide testing results?

Thank you in advance for your answers!

**Comment by kshakir** [comment]
> Does it mean that the credentials, which Travis and Jenkins use, are default credentials and the purpose of this task is to set other ones and check that four tests were successfully passed?

Using "default credentials" means that any one of a very long list of configuration setups are auto-detected by the AWS Java SDK. In each SDK-configuration case, the credentials are **not** read from the cromwell-config file and then the values passed on to the AWS SDK. Instead the SDK "discovers" them. https://docs.aws.amazon.com/sdk-for-java/v2/developer-guide/credentials.html#credentials-default

This ticket is to instead wire in credentials to the SDK via [Explicitly Specifying Credentials](https://docs.aws.amazon.com/sdk-for-java/v2/developer-guide/credentials.html#credentials-explicit).

In the tests:
- Each of the "[Default Credential Provider](https://docs.aws.amazon.com/sdk-for-java/v2/developer-guide/credentials.html#credentials-default)" options would not be configured on the system
- The `java -Dconfig.file=..." would contain the AWS key, secret and probably region
- When the various AWS SDK functions run, they each use the passed in credentials to run the commands for S3, Batch, etc.
- The jobs should still run to completion

**Comment by senin26** [comment]
Hello @kshakir,
I'm also interested in solving of this issue, could you please confirm that the expected solution implies that **some functionality should be added to satisfy the Explicitly specifying credentials**? My proposal is to write **some code that process the user credentials from the properties-file instead of using default credentials**. Does it make sense?

**Comment by likeanowl** [comment]
Hi @kshakir! I just joined @senin26 on this issue solving. So, your answer on previous question is really important to us. 
Hope for soon response! :)

**Comment by senin26** [comment]
Hello @kshakir, if you give us some clarifications, we will really appreciate it

**Comment by kshakir** [comment]
Sorry for the delay. Our organization moved our issue tracking over to Jira last month.

https://github.com/broadinstitute/cromwell/blob/43/README.md#issue-tracking-is-now-on-jira

Let's continue the active ticket there:

https://broadworkbench.atlassian.net/browse/BA-4740

I'll close this old issue here.
